src/store.ts

```diff
@@ -0,0 +1,921 @@
+    ) {
+      record = {
+        id: record.id,
+        state: to,
+        timeout: record.timeout,
+        iKeyForCreate: record.iKeyForCreate,
+        iKeyForComplete: ikey,
+        param: record.param,
+        value: value,
+        tags: record.tags,
+        createdOn: record.createdOn,
+        completedOn: time,
+        callbacks: record.callbacks,
+      };
+
+      this.promises.set(id, record);
+      return { promise: record, applied: true };
+    }
+
+    if (
+      record?.state === "pending" &&
+      ["resolved", "rejected", "rejected_canceled"].includes(to) &&
+      !strict &&
+      time >= record.timeout
+    ) {
+      return this.transition(id, "rejected_timedout");
+    }
+
+    if (
+      record?.state === "pending" &&
+      ["resolved", "rejected", "rejected_canceled"].includes(to) &&
+      strict &&
+      time >= record.timeout
+    ) {
+      throw new Error("the promise is already timeout");
+    }
+
+    if (record?.state === "pending" && to === "rejected_timedout") {
+      util.assert(time >= record.timeout);
+
+      record = {
+        id: record.id,
+        state: record.tags?.["resonate:timeout"] === "true" ? "resolved" : to,
+        timeout: record.timeout,
+        iKeyForCreate: record.iKeyForCreate,
+        param: record.param,
+        value: record.value,
+        tags: record.tags,
+        createdOn: record.createdOn,
+        completedOn: record.completedOn,
+        callbacks: record.callbacks,
+      };
+
+      this.promises.set(id, record);
+      return { promise: record, applied: true };
+    }
+
+    if (
+      record?.state !== undefined &&
+      [
+        "resolved",
+        "rejected",
+        "rejected_canceled",
+        "rejected_timedout",
+      ].includes(record.state) &&
+      to === "pending" &&
+      !strict &&
+      ikeyMatch(record.iKeyForCreate, ikey)
+    ) {
+      return { promise: record, applied: false };
+    }
+
+    if (
+      record !== undefined &&
+      ["resolved", "rejected", "rejected_canceled"].includes(record.state) &&
+      ["resolved", "rejected", "rejected_canceled"].includes(to) &&
+      !strict &&
+      ikeyMatch(record.iKeyForComplete, ikey)
+    ) {
+      return { promise: record, applied: false };
+    }
+
+    if (
+      record !== undefined &&
+      record.state === "rejected_timedout" &&
+      ["resolved", "rejected", "rejected_canceled"].includes(to) &&
+      !strict
+    ) {
+      return { promise: record, applied: false };
+    }
+
+    if (
+      record !== undefined &&
+      ["resolved", "rejected", "rejected_canceled"].includes(record.state) &&
+      ["resolved", "rejected", "rejected_canceled"].includes(to) &&
+      strict &&
+      ikeyMatch(record.iKeyForComplete, ikey) &&
+      record.state === to
+    ) {
+      return { promise: record, applied: false };
+    }
+
+    throw new Error("unexpected transition");
+  }
+}
+
+function ikeyMatch(left?: string, right?: string): boolean {
```

Comment:
left and right in this case are not optional arguments, but arguments that can take undefined as their values. The difference is subtle, but optional arguments allow you to not pass that argument at the call site. Given that this is a comparison and you will always pass both arguments at the call site, I will probably write this function as:
```suggestion
function ikeyMatch(left: string | undefined, right: string | undefined): boolean {
```